cpu/nrf52/periph/pwm.c

```diff
@@ -90,9 +90,6 @@ uint32_t pwm_init(pwm_t pwm, pwm_mode_t mode, uint32_t freq, uint16_t res)
 
     /* pin configuration */
     for (unsigned i = 0; i < PWM_CHANNELS; i++) {
-        if (pwm_config[pwm].pin[i] != GPIO_UNDEF) {
-            NRF_P0->PIN_CNF[pwm_config[pwm].pin[i]] = PIN_CNF_SET;
-        }
         /* either left aligned pol or inverted duty cycle */
         pwm_seq[pwm][i] = (POL_MASK & mode) ? POL_MASK : res;
         dev(pwm)->PSEL.OUT[i] = pwm_config[pwm].pin[i];
```

Comment:
I take setting `OUT` to `0xFF` is the right thing to do then if we don't want to use it?

Reply:
Reading the data sheet, I think we should set it to `UINT32_MAX` (or just `1 << 31`) to properly disconnect the output in the `GPIO_UNDEF` case.

![image](https://user-images.githubusercontent.com/1301112/94675579-3b11fe00-031a-11eb-935d-8f7becf1fceb.png)

Reply:
GPIO_UNDEF, in its 0xffffffff definition, has just the right property of having a bit 1 in the 'C' field...

Just that -- and until now I just missed that -- on this platform our gpio_t is uint8_t and 0xff is the undefined value, so we've all been driving P1.31 unless the hardware caught that.

The fix I propose and will send soon is to cast the uint8_t up to the uint32_t via int8_t and int32_t, thereby getting sign extension that cheaply gets us the 0xffffffff that's *actually* disabling the output.

Reply:
Oh, cross-post -- yes, we should, and patch is on the way.

Reply:
> The fix I propose and will send soon is to cast the uint8_t up to the uint32_t via int8_t and int32_t, thereby getting sign extension that cheaply gets us the 0xffffffff that's actually disabling the output.

🤯

Reply:
Granted, [the savings](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,j:1,lang:___c,selection:(endColumn:1,endLineNumber:4,positionColumn:1,positionLineNumber:4,selectionStartColumn:1,selectionStartLineNumber:3,startColumn:1,startLineNumber:3),source:'%23include+%3Cstdint.h%3E%0A%0Astatic+volatile+uint32_t+*out+%3D+42%3B%0A%0Avoid+extend(uint8_t+pin)%0A%7B%0A++++if+(pin+%3D%3D+0xff)+%7B%0A++++++++*out+%3D+0xffffffff%3B%0A++++%7D+else+%7B%0A++++++++*out+%3D+pin%3B%0A++++%7D%0A%7D%0A%0Avoid+extend2(uint8_t+pin)%0A%7B%0A++++*out+%3D+(int32_t)(int8_t)pin%3B%0A%7D'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:armv7-cclang-trunk,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'1',trim:'1'),fontScale:14,j:1,lang:___c,libs:!(),options:'-O3',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'armv7-a+clang+(trunk)+(Editor+%231,+Compiler+%231)+C',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,wrap:'1'),l:'5',n:'0',o:'%231+with+armv7-a+clang+(trunk)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) are not as great as I had hoped, as the comparison operation can be acted on without further branching.